File: src/primo_mcp_server/formatter.py
"""Format Primo records into compact, LLM-friendly text output."""

from __future__ import annotations

from primo_mcp_server.models import PrimoRecord, SearchResponse


def _format_authors(creators: list[str], max_authors: int = 3) -> str:
    """Format an author list, truncating with 'et al.' if needed."""
    if not creators:
        return "Unknown author"
    if len(creators) <= max_authors:
        return "; ".join(creators)
    return "; ".join(creators[:max_authors]) + " et al."


def _format_identifiers(record: PrimoRecord) -> str:
    """Format the most useful identifier for a record."""
    parts = []
    if record.doi:
        parts.append(f"DOI: {record.doi}")
    if record.isbn:
        parts.append(f"ISBN: {record.isbn[0]}")
    if record.issn:
        parts.append(f"ISSN: {record.issn[0]}")
    return " | ".join(parts) if parts else ""


def _format_availability(record: PrimoRecord) -> str:
    """Format availability information."""
    parts = []
    if record.fulltext_available:
        parts.append("Full text available")
    if record.delivery_category:
        parts.append(record.delivery_category)
    return " | ".join(parts) if parts else "Check availability in OneSearch"


def _format_holdings(record: PrimoRecord) -> list[str]:
    """Physical holdings, one indented line each.

    Each line is library, call number, shelf location, and status, joined with
    " | " (empty parts dropped, library code omitted since the library name
    carries it). Returns an empty list when the record has none, which is the
    usual case for records that come from a brief search.
    """
    if not record.holdings:
        return []
    lines = ["Holdings:"]
    for h in record.holdings:
        parts = []
        if h.library:
            parts.append(h.library)
        if h.call_number:
            parts.append(h.call_number)
        if h.location:
            parts.append(h.location)
        if h.availability_status:
            parts.append(h.availability_status)
        lines.append(f"  - {' | '.join(parts)}")
    return lines


def format_search_results(response: SearchResponse, query: str, offset: int = 0) -> str:
    """Format search results as a compact numbered list.

    Each result is 3-5 lines: title, authors, metadata, identifiers, availability.
    """
    if not response.records:
        return (
            f'No results found for "{query}".\n\n'
            "Suggestions:\n"
            "- Broaden your search terms\n"
            "- Check spelling\n"
            "- Try a different search field (title, creator, subject)\n"
            "- Remove filters (resource type, date range)"
        )

    total = f"{response.info.total:,}"
    showing_start = offset + 1
    showing_end = offset + len(response.records)

    lines = [
        f'Found {total} results for "{query}" (showing {showing_start}-{showing_end})',
        "",
    ]

    for i, record in enumerate(response.records, start=showing_start):
        # Line 1: number, title, year, type
        type_badge = record.resource_type.replace("_", " ").title() if record.resource_type else "Unknown"
        year = record.creation_date[:4] if record.creation_date else "n.d."

        lines.append(f"[{i}] {record.title}")
        lines.append(f"    {_format_authors(record.creators)} | {year} | {type_badge}")

        # Line 3: journal/source + identifiers
        source_parts = []
        if record.journal_title:
            journal_info = record.journal_title
            if record.volume:
                journal_info += f", {record.volume}"
            if record.issue:
                journal_info += f"({record.issue})"
            if record.start_page:
                journal_info += f", pp. {record.start_page}"
                if record.end_page:
                    journal_info += f"-{record.end_page}"
            source_parts.append(journal_info)
        elif record.publisher:
            source_parts.append(record.publisher)

        ident = _format_identifiers(record)
        if ident:
            source_parts.append(ident)
        if source_parts:
            lines.append(f"    {' | '.join(source_parts)}")

        # Line 4: availability + peer review + record ID
        status_parts = []
        if record.peer_reviewed:
            status_parts.append("Peer-reviewed")
        status_parts.append(_format_availability(record))
        lines.append(f"    {' | '.join(status_parts)}")
        lines.append(f"    Record ID: {record.record_id}")
        lines.append("")

    return "\n".join(lines).rstrip()


def format_record_detail(record: PrimoRecord) -> str:
    """Format a single record with full details."""
    lines = []

    lines.append(f"Title: {record.title}")
    lines.append(f"Author(s): {_format_authors(record.creators, max_authors=10)}")

    if record.contributors:
        lines.append(f"Contributor(s): {'; '.join(record.contributors)}")

    year = record.creation_date[:4] if record.creation_date else "n.d."
    lines.append(f"Year: {year}")
    lines.append(f"Type: {record.resource_type.replace('_', ' ').title() if record.resource_type else 'Unknown'}")

    if record.publisher:
        lines.append(f"Publisher: {record.publisher}")

    if record.journal_title:
        journal = record.journal_title
        if record.volume:
            journal += f", vol. {record.volume}"
        if record.issue:
            journal += f", no. {record.issue}"
        if record.start_page:
            journal += f", pp. {record.start_page}"
            if record.end_page:
                journal += f"-{record.end_page}"
        lines.append(f"Journal: {journal}")

    if record.language:
        lines.append(f"Language: {record.language}")

    # Identifiers
    if record.doi:
        lines.append(f"DOI: {record.doi}")
    if record.isbn:
        lines.append(f"ISBN: {', '.join(record.isbn)}")
    if record.issn:
        lines.append(f"ISSN: {', '.join(record.issn)}")

    if record.subjects:
        lines.append(f"Subjects: {'; '.join(record.subjects)}")
    if record.keywords:
        lines.append(f"Keywords: {'; '.join(record.keywords)}")

    lines.append(f"Peer-reviewed: {'Yes' if record.peer_reviewed else 'No'}")

    if record.description:
        # Truncate long descriptions
        desc = record.description
        if len(desc) > 500:
            desc = desc[:497] + "..."
        lines.append(f"\nDescription:\n{desc}")

    # Availability
    lines.append(f"\nAvailability: {_format_availability(record)}")
    lines.extend(_format_holdings(record))
    if record.source_label:
        lines.append(f"Source: {record.source_label}")

    lines.append(f"Record ID: {record.record_id}")

    return "\n".join(lines)


def format_suggestions(suggestions: list[str], query: str) -> str:
    """Format autocomplete suggestions."""
    if not suggestions:
        return f'No suggestions found for "{query}".'

    lines = [f'Suggestions for "{query}":', ""]
    for s in suggestions:
        lines.append(f"  - {s}")
    return "\n".join(lines)

File: src/formatter.ts
/**
 * Format Primo records into compact, LLM-friendly text (port of formatter.py).
 *
 * The availability fallback uses the configured discovery name (e.g. "PittCat",
 * "OneSearch") rather than a hardcoded brand, so the caller passes it in.
 */
import type { PrimoRecord, SearchResponse } from "./models.js";

/** Title-case a space-separated string (mirrors Python str.title() for these inputs). */
function titleCase(value: string): string {
  return value
    .split(" ")
    .map((w) => (w.length > 0 ? w.charAt(0).toUpperCase() + w.slice(1).toLowerCase() : w))
    .join(" ");
}

function typeBadge(resourceType: string): string {
  return resourceType ? titleCase(resourceType.replace(/_/g, " ")) : "Unknown";
}

function yearOf(creationDate: string): string {
  return creationDate ? creationDate.slice(0, 4) : "n.d.";
}

function formatAuthors(creators: string[], maxAuthors = 3): string {
  if (creators.length === 0) return "Unknown author";
  if (creators.length <= maxAuthors) return creators.join("; ");
  return creators.slice(0, maxAuthors).join("; ") + " et al.";
}

function formatIdentifiers(record: PrimoRecord): string {
  const parts: string[] = [];
  if (record.doi) parts.push(`DOI: ${record.doi}`);
  if (record.isbn.length > 0) parts.push(`ISBN: ${record.isbn[0]}`);
  if (record.issn.length > 0) parts.push(`ISSN: ${record.issn[0]}`);
  return parts.length > 0 ? parts.join(" | ") : "";
}

function formatAvailability(record: PrimoRecord, discoveryName: string): string {
  const parts: string[] = [];
  if (record.fulltextAvailable) parts.push("Full text available");
  if (record.deliveryCategory) parts.push(record.deliveryCategory);
  return parts.length > 0
    ? parts.join(" | ")
    : `Check availability in ${discoveryName}`;
}

export function formatSearchResults(
  response: SearchResponse,
  query: string,
  offset: number,
  discoveryName: string,
): string {
  if (response.records.length === 0) {
    return (
      `No results found for "${query}".\n\n` +
      "Suggestions:\n" +
      "- Broaden your search terms\n" +
      "- Check spelling\n" +
      "- Try a different search field (title, creator, subject)\n" +
      "- Remove filters (resource type, date range)"
    );
  }

  const total = response.info.total.toLocaleString("en-US");
  const showingStart = offset + 1;
  const showingEnd = offset + response.records.length;

  const lines: string[] = [
    `Found ${total} results for "${query}" (showing ${showingStart}-${showingEnd})`,
    "",
  ];

  let index = showingStart;
  for (const record of response.records) {
    lines.push(`[${index}] ${record.title}`);
    lines.push(
      `    ${formatAuthors(record.creators)} | ${yearOf(record.creationDate)} | ${typeBadge(record.resourceType)}`,
    );

    const sourceParts: string[] = [];
    if (record.journalTitle) {
      let journalInfo = record.journalTitle;
      if (record.volume) journalInfo += `, ${record.volume}`;
      if (record.issue) journalInfo += `(${record.issue})`;
      if (record.startPage) {
        journalInfo += `, pp. ${record.startPage}`;
        if (record.endPage) journalInfo += `-${record.endPage}`;
      }
      sourceParts.push(journalInfo);
    } else if (record.publisher) {
      sourceParts.push(record.publisher);
    }
    const ident = formatIdentifiers(record);
    if (ident) sourceParts.push(ident);
    if (sourceParts.length > 0) lines.push(`    ${sourceParts.join(" | ")}`);

    const statusParts: string[] = [];
    if (record.peerReviewed) statusParts.push("Peer-reviewed");
    statusParts.push(formatAvailability(record, discoveryName));
    lines.push(`    ${statusParts.join(" | ")}`);
    lines.push(`    Record ID: ${record.recordId}`);
    lines.push("");
    index += 1;
  }

  return lines.join("\n").replace(/\s+$/, "");
}

export function formatRecordDetail(
  record: PrimoRecord,
  discoveryName: string,
): string {
  const lines: string[] = [];

  lines.push(`Title: ${record.title}`);
  lines.push(`Author(s): ${formatAuthors(record.creators, 10)}`);
  if (record.contributors.length > 0) {
    lines.push(`Contributor(s): ${record.contributors.join("; ")}`);
  }
  lines.push(`Year: ${yearOf(record.creationDate)}`);
  lines.push(`Type: ${typeBadge(record.resourceType)}`);

  if (record.publisher) lines.push(`Publisher: ${record.publisher}`);

  if (record.journalTitle) {
    let journal = record.journalTitle;
    if (record.volume) journal += `, vol. ${record.volume}`;
    if (record.issue) journal += `, no. ${record.issue}`;
    if (record.startPage) {
      journal += `, pp. ${record.startPage}`;
      if (record.endPage) journal += `-${record.endPage}`;
    }
    lines.push(`Journal: ${journal}`);
  }

  if (record.language) lines.push(`Language: ${record.language}`);
  if (record.doi) lines.push(`DOI: ${record.doi}`);
  if (record.isbn.length > 0) lines.push(`ISBN: ${record.isbn.join(", ")}`);
  if (record.issn.length > 0) lines.push(`ISSN: ${record.issn.join(", ")}`);
  if (record.subjects.length > 0) lines.push(`Subjects: ${record.subjects.join("; ")}`);
  if (record.keywords.length > 0) lines.push(`Keywords: ${record.keywords.join("; ")}`);

  lines.push(`Peer-reviewed: ${record.peerReviewed ? "Yes" : "No"}`);

  if (record.description) {
    let desc = record.description;
    if (desc.length > 500) desc = desc.slice(0, 497) + "...";
    lines.push(`\nDescription:\n${desc}`);
  }

  lines.push(`\nAvailability: ${formatAvailability(record, discoveryName)}`);
  if (record.sourceLabel) lines.push(`Source: ${record.sourceLabel}`);
  lines.push(`Record ID: ${record.recordId}`);

  return lines.join("\n");
}

export function formatSuggestions(suggestions: string[], query: string): string {
  if (suggestions.length === 0) return `No suggestions found for "${query}".`;
  const lines: string[] = [`Suggestions for "${query}":`, ""];
  for (const s of suggestions) lines.push(`  - ${s}`);
  return lines.join("\n");
}
